### PR TITLE
feat: Edit the cap in product details page

### PR DESCRIPTION
**Pull request by NitishAroraKPMG**
## Description

Allow operator(s) to edit the caps.

## Testing instructions

Create a capped ticket and edit the caps in product details page.

### CODE REVIEW

File: repos/fdbt-site/src/pages/products/productDetails.tsx [hunk 1]
```diff
@@ -366,6 +367,17 @@ const createProductDetails = async (
         productDetailsElements.push({ id: 'time-restriction', name: 'Only valid during term time', content: ['Yes'] });
     }
 
+    if ('cap' in ticket && ticket.cap) {
```
Comment (alainajaffer) [inline repos/fdbt-site/src/pages/products/productDetails.tsx:370]: what if we have no caps but wanted to add some? I think this check should see if we have already created caps.

File: repos/fdbt-site/src/pages/products/productDetails.tsx [hunk 2]
```diff
@@ -366,6 +367,17 @@ const createProductDetails = async (
...
+        const cap = (await getCapByNocAndId(noc, ticket.cap.id)) as Cap;
+
+        productDetailsElements.push({
+            id: 'cap',
+            name: 'Cap',
+            content: [`${sentenceCaseString(cap.capDetails.name)} - £${cap.capDetails.price}`],
```
Comment (alainajaffer) [inline repos/fdbt-site/src/pages/products/productDetails.tsx:376]: If we have created caps then it is possible we didn't assign any so we should have and N/A option

File: repos/fdbt-site/src/pages/products/productDetails.tsx (hunk 1 above)
Comment (NitishAroraKPMG) [reply repos/fdbt-site/src/pages/products/productDetails.tsx:370]: this is for the edit functionality and do we need to show the N/A

File: repos/fdbt-site/src/pages/api/selectCaps.ts [hunk 3]
```diff
@@ -47,19 +46,42 @@ export default async (req: NextApiRequestWithSession, res: NextApiResponse): Pro
...
 
-        if (!Number.isInteger(Number(cap))) {
+        if (capChoice === 'yes' && !Number.isInteger(Number(cap))) {
             throw Error(`Received invalid cap id: ${cap}`);
         }
 
-        const selectedCap = await getCapContent(Number(cap), noc);
+        const selectedCap = capChoice === 'yes' ? await getCapContent(Number(cap), noc) : undefined;
+        const ticket = getSessionAttribute(req, MATCHING_JSON_ATTRIBUTE);
+        const matchingJsonMetaData = getSessionAttribute(req, MATCHING_JSON_META_DATA_ATTRIBUTE);
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/api/selectCaps.ts:55]: If we move these up to the top of the file, we can keep the `if (capChoice === 'no') {` check and just include `&& !(ticket && matchingJsonMetaData)`

File: repos/fdbt-site/src/pages/api/selectCaps.ts [hunk 4]
```diff
@@ -47,19 +46,42 @@ export default async (req: NextApiRequestWithSession, res: NextApiResponse): Pro
...
+                ...ticket,
+                cap: capChoice === 'yes' && !!selectedCap ? { id: selectedCap.id } : undefined,
+            };
+
+            await putUserDataInProductsBucketWithFilePath(updatedTicket, matchingJsonMetaData.matchingJsonLink);
+            updateSessionAttribute(req, CAPS_DEFINITION_ATTRIBUTE, undefined);
 
-        if (capChoice === 'yes' && !!selectedCap) {
-            updateSessionAttribute(req, CAPS_DEFINITION_ATTRIBUTE, {
-                id: selectedCap.id,
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/api/selectCaps.ts:58]: Can we put the file back to how it was structured before, and just move the edit mode check higher up - this means we dont need to edit the flow so much.

File: repos/fdbt-site/src/pages/products/productDetails.tsx (hunk 1 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/products/productDetails.tsx:370]: I agree with Alaina - if you look at the salesConfirmation.tsx file, we have 

`  if (['single', 'return', 'flatFare'].includes(fareType) && hasCaps) {
        confirmationElements.push({
            name: 'Cap',
            content: selectedCap?.capDetails.name || 'N/A',
            href: 'selectCaps',
        });
    }`

So we need the same thing inside the if statement here

File: repos/fdbt-site/src/pages/products/productDetails.tsx (hunk 2 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/products/productDetails.tsx:376]: Agree with Alaina

File: repos/fdbt-site/src/pages/products/productDetails.tsx (hunk 1 above)
Comment (RobCattonKPMG) [reply repos/fdbt-site/src/pages/products/productDetails.tsx:370]: Can we also only display this on test + local environment

File: repos/fdbt-site/src/utils/apiUtils/index.ts [hunk 5]
```diff
@@ -270,3 +270,7 @@ export const dateIsOverThirtyMinutesAgo = (inputDate: Date): boolean => {
 export const isADayOfTheWeek = (input: string | undefined): boolean => {
     return !!input && daysOfWeek.includes(input);
 };
+
+export const isCapTicket = (fareType: string): boolean => {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/utils/apiUtils/index.ts:274]: Name needs changing to 'fareTypeIsAllowedToAddACap'

File: repos/fdbt-site/src/pages/products/productDetails.tsx [hunk 6]
```diff
@@ -366,6 +369,23 @@ const createProductDetails = async (
         productDetailsElements.push({ id: 'time-restriction', name: 'Only valid during term time', content: ['Yes'] });
     }
 
+    const hasCaps = (await getCaps(noc)).length > 0;
+    if (isCapTicket(ticket.type) && hasCaps) {
+        let capContent = 'N/A';
+
+        if ('cap' in ticket && ticket.cap) {
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/products/productDetails.tsx:376]:
```suggestion
        if (ticket.cap) {
```

File: repos/fdbt-site/src/pages/products/productDetails.tsx [hunk 7]
```diff
@@ -366,6 +369,23 @@ const createProductDetails = async (
...
+            const cap = (await getCapByNocAndId(noc, ticket.cap.id)) as Cap;
+            capContent = `${sentenceCaseString(cap.capDetails.name)} - £${cap.capDetails.price}`;
+        }
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/products/productDetails.tsx:379]: We need to feature flag this, so it only happens on dev or test.

File: repos/fdbt-site/src/pages/products/productDetails.tsx [hunk 8]
```diff
@@ -366,6 +369,25 @@ const createProductDetails = async (
         productDetailsElements.push({ id: 'time-restriction', name: 'Only valid during term time', content: ['Yes'] });
     }
 
+    const hasCaps = (await getCaps(noc)).length > 0;
+    const isDevOrTest = process.env.NODE_ENV === 'development' || process.env.STAGE === 'test';
```
Comment (RobCattonKPMG) [inline repos/fdbt-site/src/pages/products/productDetails.tsx:373]: This cannot be called here - it must be done in the serverSideProps and passed through as a prop

**Review by RobCattonKPMG — APPROVED** [review]

**Pull request merged**